### PR TITLE
[FLINK-20482][python] Support General Python UDF for Map Operation in Python Table API

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support General Python UDF for Map Operation in Python Table API*

## Brief change log

  - *Add rewrite rule `PythonCalcSplitRexFieldRule`*
  - *Add `map` operation in Python `Table`*

## Verifying this change

This change added tests and can be verified as follows:

  - *Add UT in `PythonCalcSplitRuleTest`*
  - *Add IT in `test_row_based_operation.py`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PythonCalcSplitRule.scala [hunk 1]
```diff
@@ -333,6 +387,18 @@ private class ScalarFunctionSplitter(
     extractedRexNodes.append(node)
     newNode
   }
+
+  private def getExtractRexFieldNode(node: RexFieldAccess, rexCallIndex: Int): RexNode = {
+    val pythonCall: RexCall = program.expandLocalRef(
+      node.getReferenceExpr.asInstanceOf[RexLocalRef]).asInstanceOf[RexCall]
+    if (!fieldsRexCall.contains(rexCallIndex)) {
+      extractedRexNodes.append(pythonCall)
+      fieldsRexCall += rexCallIndex -> (extractedFunctionOffset + extractedRexNodes.length -1)
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PythonCalcSplitRule.scala:396]:
```suggestion
      fieldsRexCall += rexCallIndex -> (extractedFunctionOffset + extractedRexNodes.length - 1)
```

File: flink-python/pyflink/table/tests/test_row_based_operation.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        table_sink = source_sink_utils.TestAppendSink(
+            ['a', 'b'],
+            [DataTypes.BIGINT(), DataTypes.BIGINT()])
+        self.t_env.register_table_sink("Results", table_sink)
+
+        func = udf(lambda x: Row(x + 1, x * x), result_type=DataTypes.ROW(
+            [DataTypes.FIELD("a", DataTypes.BIGINT()),
+             DataTypes.FIELD("b", DataTypes.BIGINT())]))
+
+        t.map(func(t.b)).alias("a", "b").execute_insert("Results").wait()
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_row_based_operation.py:45]: What about adding the following test cases?
- test case for two continuous map
- test case for pandas udf

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks a lot for the PR. Have left a few comments.

File: flink-python/pyflink/table/table.py [hunk 3]
```diff
@@ -759,6 +759,27 @@ def drop_columns(self, *fields: Union[str, Expression]) -> 'Table':
...
+        Performs a map operation with an user-defined scalar function.
+
+        Example:
+        ::
+
+            >>> add = udf(lambda x: Row(x + 1, x *x), result_type=DataTypes.Row(
+            ... [DataTypes.FIELD("a", DataTypes.INT()), DataTypes.FIELD("b", DataTypes.INT())]))
+            >>> table_env.create_temporary_function("add", add)
+            >>> tab.map(add(tab.a)).alias("a, b")
+            >>> tab.map("add(a)").alias("a, b")
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/table/table.py:773]: Yes. Make sense.

File: flink-python/pyflink/table/tests/test_row_based_operation.py (hunk 2 above)
Comment (HuangXingBo) [reply flink-python/pyflink/table/tests/test_row_based_operation.py:45]: Supporting Pandas UDF will be in [FLINK-20483](https://issues.apache.org/jira/browse/FLINK-20483).
I will add the test case for two continuous map.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

**Review by dianfu — APPROVED** [review]
LGTM. +1

**Pull request closed without merging**